### ISSUE TITLE
Angle constraints

### ISSUE THREAD

**Issue by automata** [liabru/matter-js#71]
Distance constraints are really nice, how about angle constraints?

That would make it possible to remove those two "diagonal constraints" from this example: http://jsbin.com/risahi/4/edit

**Comment by liabru** [comment]
This is definitely on my list, I'll keep this thread updated, thanks!

**Comment by gamecubate** [comment]
+1

**Comment by Misiur** [comment]
+1

**Comment by jstoeffler** [comment]
+1

**Comment by iamjoshua** [comment]
+1

**Comment by iamjoshua** [comment]
What are/were your thoughts on the best way to implement this @liabru? I really need this and would be willing to help.

**Comment by davearel** [comment]
+1

**Comment by greenSnot** [comment]
+1111111111

**Comment by babyzone2004** [comment]
+1

**Comment by omegahacker** [comment]
+1

**Comment by Twosies** [comment]
Is there any progress on this one?  I need this for my project so much that I'm looking into hacking it in myself.  But, I'm concerned I might not figure out how to do it in a way that conserves momentum, etc.

**Comment by forresto** [comment]
FWIW, here's how I implemented angular constraints for another verlet lib: https://github.com/softfab/tshirt/blob/master/src/verlet-constraint-angle-2d.js

**Comment by AndrewBrownK** [comment]
I will bury a timecapsule with instructions for my great grandchildren on how to complete my project when angle constraints are finally added to this library in 2050

**Comment by davearel** [comment]
Anyone wanna help @liabru out? It's an open source library, not a paid service ;)

**Comment by joel-simon** [comment]
+1

**Comment by flukeout** [comment]
+1

**Comment by SillyFreak** [comment]
My use case was to implement a servo arm. What I ended up doing was to add two revolute constraints; one for the pivot point (like flukeout) and one farther out (unlike in flukeout's example, this is also a revolute constraint!). The constraint can be modified at runtime to change to which angle the dependent element should go.

```js
const anchor = Matter.Bodies.rectangle(200, 350, 50, 50, {
  density: 1,
  frictionAir: 0.4,
});
const arm = Matter.Bodies.rectangle(200, 200, 10, 200, {
  // in this example I don't want the anchor to rotate too much when the arm moves
  density: 0.01,
  frictionAir: 0.1,
});
// this constraint remains constant
const pivot = Matter.Constraint.create({
  bodyA: anchor,
  pointA: { x: 0, y: -50 },
  bodyB: arm,
  pointB: { x: 0, y: 100 },
});
const control = Matter.Constraint.create({
  bodyA: anchor,
  // the pointA is modified to change the point to which the arm should be fixed
  pointA: { x: 0, y: -100 },
  bodyB: arm,
  pointB: { x: 0, y: 50 },
});

Matter.World.add(this.simulation.world, [anchor, arm, pivot, control]);

setTimeout(() => {
  // change the control point later
  // should use sine and cosine, this is fine for the demo
  control.pointA = { x: 20, y: -100 };
}, 1000);
```

I think by making the `pivot` constraint stiffer than the `control` constraint, you can ensure that the rotation point is at/closer to the pivot, and by adjusting the distance between pivot and control, you can control the "angular stiffness".

**Comment by liabru** [comment]
I've opened PR #837 which adds angle constraints, those interested please take a look, try them out and let me know if you spot any issues or just if they work nicely for your use case.

Thanks for holding on here everyone!

(@AndrewBrownK get ready to dig up that time capsule)

**Comment by ulfaslak** [comment]
Phew! I really needed this feature. Was quite exciting reading through the comments starting 6 years ago and finding that it had finally happened by 2020. Good job @liabru !

**Comment by liabru** [comment]
Yes! I need to come back to this though, it's not merged as there were a few cases that were not stable (chains). I could probably release it with some caveats I suppose.

**Comment by joel-simon** [comment]
@liabru any updates on this? Thanks!

**Comment by mreinstein** [comment]
I'm trying to simulate some piston-like things and this angle constraint would make things significantly easier.
